nemo_text_processing/g2p/modules.py

```diff
@@ -404,39 +478,54 @@ def _normalize_dict(self, phoneme_dict_obj: dict) -> Tuple[DefaultDict[str, List
         g2p_dict = defaultdict(list)
         symbols = set()
         for word, prons in phoneme_dict_obj.items():
+            # process word
             # update word cases.
-            if self.set_graphemes_upper:
-                word_new = word.upper()
-            else:
-                word_new = word.lower()
+            word_new = set_grapheme_case(word, case=self.grapheme_case)
 
             # add grapheme symbols if `use_chars=True`.
             if self.use_chars:
-                # remove punctuations after a word.
+                # remove punctuations within a word. Punctuations can exist at the start, middle, and end of a word.
+                # TODO @xueyang: is the removal of a hyphen within a word expected, e.g. up-to-date -> uptodate?
                 word_no_punct = self.PUNCT_REGEX.sub('', word_new)
-                symbols.update(word_no_punct)
 
+                # add prefix to distinguish graphemes from phonemes.
+                symbols.update(self._prepend_prefix_for_one_word(word_no_punct))
+
+            # process IPA pronunciations
             # update phoneme symbols by removing lexical stress markers if `use_stresses=False`.
             prons_new = list()
             if not self.use_stresses:
                 for pron in prons:
-                    pron_no_stress = [symbol for symbol in pron if symbol not in self.STRESS_SYMBOLS]
-                    prons_new.append(pron_no_stress)
-                    symbols.update(pron_no_stress)
+                    prons_new.append([symbol for symbol in pron if symbol not in self.STRESS_SYMBOLS])
             else:
                 prons_new = prons
-                for pron in prons:
-                    symbols.update(pron)  # This will insert each char individually
 
-            # update final dict
+            # update symbols
+            for pron in prons_new:
+                symbols.update(pron)  # This will insert each char individually
+
+            # update dict entry
             g2p_dict[word_new] = prons_new
 
+            # duplicate word entries if grapheme_case is mixed. Even though grapheme_case is set to mixed, the words in
+            # the original input text and the g2p_dict remain unchanged, so they could still be either lowercase,
+            # uppercase, or mixed-case as defined in `set_grapheme_case` func. When mapping an uppercase word, e.g.
+            # "HELLO", into phonemes using the g2p_dict with {"Hello": [["həˈɫoʊ"]]}, "HELLO" can't find its
+            # pronunciations in the g2p_dict due to the case-mismatch of the words. Augmenting the g2p_dict with its
+            # uppercase word entry, e.g. {"Hello": [["həˈɫoʊ"]], "HELLO": [["həˈɫoʊ"]]} would provide possibility to
+            # find "HELLO"'s pronunciations rather than directly considering it as an OOV.
+            if self.grapheme_case == GRAPHEME_CASE_MIXED and not word_new.isupper():
+                g2p_dict[word_new.upper()] = prons_new
```

Comment:
Why not just set them all to `.upper()` to avoid potential duplicate entries?

Reply:
this is not an expected behavior if setting all words to `upper()`. For example, if you look at German's dictionary, you will find different pronunciations between words `Weg` and `weg`.
```
Weg  vˈeːk
weg  vˈɛk
```

Reply:
Ah, I see! Didn't realize that there were cases like this. Very interesting.